File: DS/Trees/Heap/MaxHeap.java
//Very similar to MinHeap 
//Refer: https://www.geeksforgeeks.org/binary-heap/
public class MaxHeap{

	private int capacity;
	private int heapSize;
	private int[] harr;

	public MaxHeap(int cap){
		capacity = cap;
		heapSize = 0;
		harr = new int[capacity];
	}
	private int parent(int i){
		return (i-1)/2;
	}
	private int left(int i){
		return (2*i) + 1;
	}
	private int right(int i){
		return (2*i) + 2;
	}
	//Swap the value in heap array
	private void swap(int a, int b){
		int temp = harr[a];
		harr[a] = harr[b];
		harr[b] = temp;
	}
	//Prints the heap
	public void print(){
		System.out.println();
		for(int i=0; i<heapSize; i++){
			System.out.print(harr[i]+" ");
		}
	}
	////Returns the max value of the heap
	public int getMax(){
		return harr[0];
	}

	//Inserts the key into the heap
	public void insert(int key){

		if(heapSize == capacity){
			System.out.println("OverFlow: Heap is full");
		}

		//First insert the key at the end of heap
		int i = heapSize;
		heapSize++;
		harr[i] = key;
		//Check max heap property for all its ancestors
		while(i != 0 && harr[i] > harr[parent(i)]){
			swap(parent(i), i);
			i = parent(i);
		}

	}
	//Returns and removes the maximum element from the heap
	public int extractMax(){

		if(heapSize == 0)
			return Integer.MIN_VALUE;
		if(heapSize == 1){
			heapSize--;
			return harr[0];
		}

		//Store the maximum value and remove it from the heap
		int max = harr[0];

		//Put the last value of the heap into the first
		harr[0] = harr[heapSize - 1];
		heapSize--;
		maxHeapify(0);

		return max;
	}
	// A Recursive method to heapify a subtree with root at given index
	// This method assumes that the subtree are already heapified
	public void maxHeapify(int i){
		int l = left(i);
		int r = right(i);
		int largest = i;

		if(l < heapSize && harr[l] > harr[largest])
			largest = l;
		if(r < heapSize && harr[r] > harr[largest])
			largest = r;

		if(largest != i){
			swap(largest, i);
			maxHeapify(largest);
		}
	}
	// This function deletes key at index i. It first increases value to
	// infinite, then calls extractMin()
	public void deleteKey(int i){
		increaseKey(i, Integer.MAX_VALUE);
		extractMax();
	}
	// Increases value at index 'i' to newval.  It is assumed that
	// newval is greater than harr[i].
	public void increaseKey(int i, int newval){
		harr[i] = newval;
		while(i != 0 && harr[i] > harr[parent(i)]){
			swap(i, parent(i));
			i = parent(i);
		}
	}


}

File: DS/Trees/Heap/max_heap.rb
class MaxHeap
    attr_accessor :harr, :capacity, :heap_size

    def initialize(capacity)
        @heap_size = 0
        @capacity = capacity
        @harr = Array.new(@capacity)
    end

    def insert(val)
        raise 'Overflow' if @heap_size == @capacity

        i = @heap_size
        @heap_size += 1
        @harr[i] = val

        while(i != 0 && @harr[i] > @harr[parent(i)])
            @harr[parent(i)], @harr[i] = @harr[i], @harr[parent(i)]
			i = parent(i);
		end
    end

    def get_max()
        @harr[0]
    end

    def extract_max()
        raise 'empty heap' if @heap_size == 0

        if @heap_size == 1
            @heap_size -= 1
            return @harr[0]
        end

        max = @harr[0]
        @harr[0] = @harr[@heap_size - 1]
        @heap_size -= 1
        max_heapify(0)
        max
    end

    def max_heapify(i)
        l = left(i); r = right(i)
        largest = i

        largest = l if (l < @heap_size && @harr[l] > @harr[i])
        largest = r if (r < @heap_size && @harr[r] > @harr[largest])

        if (largest != i)
            @harr[largest], @harr[i] = @harr[i], @harr[largest]
            max_heapify(largest)
        end
    end

    private
        def parent(i)
            (i-1)/2
        end

        def left(i)
            (2*i) + 1
        end

        def right(i)
            (2*i) + 2
        end
end
mh = MaxHeap.new(5)
mh.insert(-2)
mh.insert(-7)
mh.insert(-2)
mh.insert(-16)
puts mh.extract_max
# mh.insert(4)
# puts mh.extract_max
# puts mh.extract_max
# puts mh.extract_max
# puts mh.extract_max
class MaxHeap
    attr_accessor :harr, :hsize, :capacity
    
    def initialize(capacity)
        @capacity = capacity
        @hsize = 0
        @harr = Array.new(@capacity)
    end
    
    def insert(val)
        return if full?
        
        i = @hsize
        @harr[@hsize] = val
        @hsize += 1
        
        while(i != 0 && @harr[parent(i)] < @harr[i])
            @harr[parent(i)], @harr[i] = @harr[i], @harr[parent(i)]
            i = parent(i)
        end
    end
    
    def extract()
        return if empty?
        
        max = peak()
        @hsize -= 1
        @harr[0] = @harr[@hsize]
        heapify(0)
        max
    end
    
    def size
        @hsize
    end
    
    def peak()
        @harr[0]
    end
    
    def empty?
        @hsize == 0
    end
    
    def full?
        @hsize == @capacity
    end
    
    private
    
    def heapify(i)
        l = left(i)
        r = right(i)
        lr = i
        
        lr = l if l < @hsize && @harr[i] < @harr[l]
        lr = r if r < @hsize && @harr[lr] < @harr[r]
        
        if (i != lr)
            @harr[i], @harr[lr] = @harr[lr], @harr[i]
            heapify(lr)
        end
    end
    
    def parent(i)
        (i-1)/2
    end

    def left(i)
        (2*i) + 1
    end

    def right(i)
        (2*i) + 2
    end
end